service/s3/s3manager/batch.go

```diff
@@ -309,6 +323,77 @@ func (d *BatchDelete) Delete(ctx aws.Context, iter BatchDeleteIterator) error {
 	return nil
 }
 
+// deleteBatch will delete a batch of items in the objects parameters.
+func deleteBatch(d *BatchDelete, objects []BatchDeleteObject) []Error {
+	errs := []Error{}
+	for len(objects) > 0 {
+		object := objects[0]
+		objects = objects[1:]
+
+		input := &s3.DeleteObjectsInput{
+			Bucket: object.Object.Bucket,
+			Delete: &s3.Delete{
+				Objects: []*s3.ObjectIdentifier{
+					{
+						Key:       object.Object.Key,
+						VersionId: object.Object.VersionId,
+					},
+				},
+			},
+			MFA:          object.Object.MFA,
+			RequestPayer: object.Object.RequestPayer,
+		}
+
+		// tempObjects will hold objects that do not match the proper group of bucket,
+		// MFA, and request payer.
+		tempObjects := []BatchDeleteObject{}
+		// afters will hold objects that need to run the cleanup function for objects
+		// that have been deleted
+		afters := []BatchDeleteObject{}
+		for i := 0; i < len(objects); i++ {
+			if object.Object.Bucket != nil && objects[i].Object.Bucket != nil {
+				if *object.Object.Bucket != *objects[i].Object.Bucket {
+					tempObjects = append(tempObjects, objects[i])
+					continue
+				}
+			}
+
+			if object.Object.MFA != nil && objects[i].Object.MFA != nil {
+				if *object.Object.MFA != *objects[i].Object.MFA {
+					tempObjects = append(tempObjects, objects[i])
+					continue
+				}
+			}
+
+			if object.Object.RequestPayer != nil && objects[i].Object.RequestPayer != nil {
+				if *object.Object.RequestPayer != *objects[i].Object.RequestPayer {
+					tempObjects = append(tempObjects, objects[i])
+					continue
+				}
+			}
+			input.Delete.Objects = append(input.Delete.Objects, &s3.ObjectIdentifier{
+				Key:       objects[i].Object.Key,
+				VersionId: objects[i].Object.VersionId,
+			})
+			afters = append(afters, objects[i])
+		}
+
+		objects = tempObjects
+		if _, err := d.Client.DeleteObjects(input); err != nil {
```

Comment:
`DeleteObjects` API will actually return successful, but with a list of errors if some of the deletes fail. http://docs.aws.amazon.com/sdk-for-go/api/service/s3/#DeleteObjectsOutput